magento/magento2
# Fix page cache controller block renderer error handling

### Description (*)
Private content handling page cache controller has poor error handling. It does not handle invalid JSON and undefined offset errors. This may not be a problem normally as this controller has limited usage but it may generate a large number of exceptions in case, for example, a security scanner generates requests with random parameters string.
* When "originalRequest" parameter has invalid JSON, the controller returns an empty array. No errors are logged.
* When "originalRequest" parameter does not have a required parameter, "Undefined index" unhandled exception is thrown.

The fix includes the next:
* Improvement to the JSON decoding. In case of an invalid JSON, an error is logged to the exception log and request forwarded to "noroute".
* Improvement to the original request parameters handling. In case of any parameter missing, the request is forwarded to "noroute".
### Manual testing scenarios (*)
1. Try to pass the parameter string with the missing required parameter for the original request.
a. Request the next URL in your browser `www.your_magento_domain/page_cache/block/render/id/111/?ajax=true&blocks=["csblock_category_content_top","csblock_category_menu_bottom","csblock_category_menu_top","csblock_category_page_bottom","csblock_category_sidebar_top","user.links.template"]&handles=["default","catalog_category_view","catalog_category_view_type_layered","catalog_category_view_id_2863","restriction_privatesales_mode"]&originalRequest={"route":"catalog","controller":"category","action":"view"}&version=df28718726589060aaf9bb4f61f5e923`
b. You'll get an exception "Notice: Undefined index: uri"
c. Expected result - an error is logged to the exception log and you get redirected to the "noroute" (404 Not Found) page.
2. Try to pass invalid JSON string in originalRequest parameter
a. Request the next URL in your browser `www.your_magento_domain/page_cache/block/render/id/111/?ajax=true&blocks=["csblock_category_content_top","csblock_category_menu_bottom","csblock_category_menu_top","csblock_category_page_bottom","csblock_category_sidebar_top","user.links.template"]&handles=["default","catalog_category_view","catalog_category_view_type_layered","catalog_category_view_id_2863","restriction_privatesales_mode"]&originalRequest={"route":"catalog","controller":"category","action"%%%&&&&:"view"}&version=df28718726589060aaf9bb4f61f5e923`
b. You'l get an empty array as a response. No errors are logged to the exception.log
c. Expected result - an error is logged to the exception log and you get redirected to the "noroute" (404 Not Found) page.

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#29665: Fix page cache controller block renderer error handling
